### ISSUE TITLE
Defining own variables in tasks.json

### ISSUE THREAD

**Issue by xaljox** [Microsoft/vscode#27829]
- VSCode Version: 1.12
- OS Version: MacOS 10.12.5 & Ubuntu 16.04

Steps to Reproduce:

I've set up a tasks.json file for building a project on multiple platforms. All platforms see the same content of the project repository. This is done either via disk sharing, because of running another platform in a VM, or via sync with the Git repository. So far so good, they all see the same task.json. However some command lines are rather long and those long lines are identical for most part. For example:

`"rm -rf build; mkdir build; cd build; ../configure --with-bash-malloc=no CFLAGS=\"-O3 -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free\" LDFLAGS=-L/usr/local/lib LIBS=\"-ltcmalloc -lcurl\" CC=clang`

Similar lines are there for the different platforms. The configure part is always the same for the different platforms, so it would be nice to factor out this common part. Thus the question is if it is possible to define your own variables, so you can use them similar to ${workspaceRoot}.

Thus define somewhere

```
"win_dir": "build_windows",
"linux_dir": "build",
"osx_dir": "build_osx",
"configure": "../configure --with-bash-malloc=no CFLAGS=\"-O3 -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free\" LDFLAGS=-L/usr/local/lib LIBS=\"-ltcmalloc -lcurl\" CC=clang"
```
And then write

```
"tasks": [
    {
        "taskName": "configure",
        "command": "bash",
        "windows": {
            "args": ["-c", "rm -rf ${win_dir}; mkdir ${win_dir}; cd ${win_dir}; ${configure}"]
        },
        "linux": {
            "args": ["-c", "rm -rf ${linux_dir}; mkdir ${linux_dir}; cd ${linux_dir}; ${configure}"]
        },
        "osx": {
            "args": ["-c", "rm -rf ${osx_dir}; mkdir ${osx_dir}; cd ${osx_dir}; ${configure}"]
        },
        "isBuildCommand": true,
        "problemMatcher": "$make-compile"
    },
    ... others tasks using the variables
```
When making changes to the build directory or arguments passed to configure etc, then the tasks.json file needs only editing at one place, instead of many.

Perhaps it is already possible but I'm unable to find out how. I tried to do something with the declares block, but that seems to be hard tied to problemMatcher. You can find some examples, but I could not find clear documentation of of the elements of the tasks.json file and how they interact.

Perhaps I'm missing something, please educate me!

While at it, some sort of macro expansion would be nice also.
that would allow e.g.
"configure_cmd_line($1)": "rm -rf ${$1}; mkdir ${$1}; cd ${$1}; ${configure}"

and the args part would then become:
```
...
"linux": {
            "args": ["-c", "configure_cmd_line(${linux_dir})"]
        },
...
```
I do not care so much about the actual syntax, as long as it is possible to factor out common used patterns that repeat themselves in the tasks.json file, especially with multi platform builds. That would make it much more easier to maintain and copy them from one project to another and only slightly modify them in at least as possible places as possible.

**Comment by dbaeumer** [comment]
@xaljox we are working on API that will allow extension to contribute tasks to the system hence giving them the full flexibility of node and JS to complete proper tasks. Instead of adding more features to the tasks.json we would like to encourage users to write extension. Would this work in your case. The API to contribute tasks is currently in proposed state and available in vscode.proposed.d.ts in our source tree. The gulp and grunt extensions are examples of the task API consumers.

**Comment by xaljox** [comment]
I’m a C/C++ programmer, trying to stay away from certain things ;-). if you simpel want to change a build directory, or a compile flag, then reprogramming an extension is not they way to go. So that extension then needs some sort of configuration file similar to the tasks.json, but with a bit more power. I seems a bit like we are now running in circles. And every extension have to revert the same wheels again.

Jimmy

**Comment by dbaeumer** [comment]
@xaljox this is a fair point. However such a support can come from a C++ extension. In general it is not our goal to provide a scripting language in the tasks.json. If we start to add variables then the next request will be control statements (if, ....). In our example wouldn't it make sense to write a shell script and then tasks simply calls the shell script? This will for example ensure that you can even compile & build without having VS Code running.

**Comment by xaljox** [comment]
It would be possible to do shell scripts I guess, but I just figured out a bit how the tasks.json stuff works and I’m almost there. Seems the next thing to figure out is how shell scripting and determining platforms works. When I look what kind of shell scripts are created with the auto tools I’m not to keen on diving into that kind of stuff. For most part the tasks.json is readable and understandable, only the link between the problemMatcher and the declares block isn’t that obvious. But when I look at some shell scripts, it really scares the hell out of you. But putting the large configure line into a small kind of bat file would certainly already help. That leaves only the directories, which can also easily be tackled with an edit of all occurances.

It is also a matter of convenience that when changing some options you only have to deal with one file. But I see your point about conditional statements etc. 

On the other side, variables are already there, for e.g. directory paths! It’s all about factoring out stuff that is repeated and there is already something for that for the problemMatcher. Thus the idea was to extend the specific solution for the problemMatcher into a more generic solution. It would then also make the problemMatcher solution with the declares less fuzzy.

The first thing I tried was to add additional stuff in the declares block, because it seems that it was doing something that I needed. But it turned out to be a specific problemMatcher thing only. I prefer to have generic solutions above specific, they tend to be easier to understand, less error prone and more versatile.

So variables are already there, their expansion in the tasks.json works fine, just a way to add a few of your own would be great, even if they are in a separate file. Or a way for an extension to add additional variables would also be a valid solution. I assume an extension can have its own GUI part, which then could be used to edit them. 

Jimmy
